### PR TITLE
Fixup of resource sections in GCC builds

### PR DESCRIPTION
**Pull request by disean**
## Proposed changes

- GCC: Make resource sections discardable for kernel-mode drivers and DLLs
- Fix pciide and pcix drivers

JIRA issue: [CORE-17401](https://jira.reactos.org/browse/CORE-17401)

### CODE REVIEW

**Comment by HBelusca** [comment]
No please don't do that. We don't need new CMake module types for each and every particular cases.
The correct fix is to keep what is currently present in master, and just add your resource section override linker flags in **your own** cmakefiles for the given module (pciide and pcix for example).

**Comment by disean** [comment]
Ok, I'll try to improve it.

File: sdk/tools/pefixup.c [hunk 1]
```diff
@@ -144,9 +144,15 @@ static int driver_fixup(int mode, unsigned char *buffer, PIMAGE_NT_HEADERS nt_he
...
-        /* For some reason, .rsrc is made writable by windres */
         if (strncasecmp((char*)Section->Name, ".rsrc", 5) == 0)
         {
+            /* .rsrc is discardable for driver images, WDM drivers, and Kernel-Mode DLLs */
+            if (mode == MODE_KERNELDRIVER || mode == MODE_WDMDRIVER || mode == MODE_KERNELDLL)
+            {
+                Section->Characteristics |= IMAGE_SCN_MEM_DISCARDABLE;
+            }
+
+            /* For some reason, .rsrc is made writable by windres */
```
Comment (HBelusca) [inline sdk/tools/pefixup.c:155]: Why does this need an extra handling there?

Comment (disean) [reply sdk/tools/pefixup.c:155]: I follow the loop logic "continue if the given section can be paged".

File: drivers/storage/ide/pciide/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,11 @@
 
 add_library(pciide MODULE pciide.c pciide.rc)
 set_module_type(pciide kernelmodedriver)
+if(MSVC)
+    target_link_options(pciide PRIVATE /SECTION:.rsrc,!D)
+else()
+    # Not implemented
```
Comment (HBelusca) [inline drivers/storage/ide/pciide/CMakeLists.txt:7]:
```suggestion
    add_custom_command(
        TARGET pciide POST_BUILD
        COMMAND objcopy --set-section-flags .rsrc=readonly $<TARGET_FILE:pciide>
        VERBATIM)
```
(Note: if "objcopy" doesn't work, try using `${CMAKE_OBJCOPY}` instead.)

File: drivers/bus/pcix/CMakeLists.txt [hunk 3]
```diff
@@ -48,6 +48,11 @@ add_library(pcix MODULE
 
 set_module_type(pcix kernelmodedriver)
 target_link_libraries(pcix arbiter)
+if(MSVC)
+    target_link_options(pcix PRIVATE /SECTION:.rsrc,!D)
+else()
+    # Not implemented
```
Comment (HBelusca) [inline drivers/bus/pcix/CMakeLists.txt:54]:
```suggestion
    add_custom_command(
        TARGET pcix POST_BUILD
        COMMAND objcopy --set-section-flags .rsrc=readonly $<TARGET_FILE:pcix>
        VERBATIM)
```
(Note: if "objcopy" doesn't work, try using `${CMAKE_OBJCOPY}` instead.)

File: sdk/tools/pefixup.c (hunk 1 above)
Comment (HBelusca) [reply sdk/tools/pefixup.c:155]: My comment was misplaced and was for the handling of the discardable flag.

Comment (disean) [reply sdk/tools/pefixup.c:155]: MSVC linker does that by default.

Comment (HBelusca) [reply sdk/tools/pefixup.c:155]: Oh sorry you were fixing also another problem in the same PR.
Ok then keep it, and adjust the comment there too: https://git.reactos.org/?p=reactos.git;a=blob;f=sdk/cmake/gcc.cmake;hb=2587d72ef7c0673a619a7ad4202a82f926e8a6ab#l300

File: drivers/storage/ide/pciide/CMakeLists.txt (hunk 2 above)
Comment (disean) [reply drivers/storage/ide/pciide/CMakeLists.txt]: objcopy is broken
result

Comment (HBelusca) [reply drivers/storage/ide/pciide/CMakeLists.txt]: huh yeah indeed definitively. It does much more changes (and sometimes harmful ones) than intended. Let's not use it then.

**Comment by HBelusca** [comment]
@disean : Before I merge your PR, could you also have a look at whether the resource section, as a generic rule for kernel drivers, is to be paged or non-paged?

**Comment by disean** [comment]
The resource section is pageable.
By the way, CORE-17401 unrelated to the pcix driver.

**Pull request merged**